src/VisualStudio/Core/Def/Implementation/EditAndContinue/DebuggeeModuleMetadataProvider.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Composition;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Debugging;
+using Microsoft.CodeAnalysis.EditAndContinue;
+using Microsoft.VisualStudio.Debugger;
+using Microsoft.VisualStudio.Debugger.Clr;
+using Microsoft.VisualStudio.Debugger.ComponentInterfaces;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.EditAndContinue
+{
+    [Export(typeof(IDebuggeeModuleMetadataProvider)), Shared]
+    internal sealed class DebuggeeModuleMetadataProvider : IDebuggeeModuleMetadataProvider
+    {
+        private sealed class DebuggerService :
+            IDkmModuleInstanceUnloadNotification,
+            IDkmProcessExitNotification,
+            IDkmCustomMessageForwardReceiver
+        {
+            public static readonly Guid ComponentId = new Guid("A96BBE03-0408-41E3-8613-6086FD494B43");
+            public static readonly Guid MessageSourceId = new Guid("58CDF976-1923-48F7-8288-B4189F5700B1");
+
+            private DebuggeeModuleMetadataProvider _provider;
```

Comment:
rather than keeping this as a member of the component, I think you need a ProviderContainer class which inherits from DkmDataItem.

In SendLower, you can create the container, and add it as a data item to DkmProcess.

In other method such as OnModuleInstanceUnload, you can get the DkmProcess from the DkmModuleInstance (probably get property RuntimeInstance.Process), get the data item from it, and call OnModuleInstanceUnload member method, which will notify the _provider it contains. Same with OnProcessExit.

Now, in the OnClose method of the container, you can call _provider.NotifyComponentUnloaded() or something, which will reset _debuggerComponentInitialized.